odoo/addons/base/models/ir_attachment.py

```diff
@@ -697,3 +711,48 @@ def action_get(self):
     def _get_serve_attachment(self, url, extra_domain=None, order=None):
         domain = [('type', '=', 'binary'), ('url', '=', url)] + (extra_domain or [])
         return self.search(domain, order=order, limit=1)
+
+
+class LinkAttachmentMixin(models.AbstractModel):
+    _name = 'link.attachment.mixin'
+    _description = "Link between attachments and records"
+
+    linked_attachment_ids = fields.One2many(
+        comodel_name='ir.attachment',
+        compute='_compute_linked_attachment_ids',
+        inverse='_inverse_linked_attachment_ids',
+    )
```

Comment:
If you use the following definition, invalidation is **automatic**:
```suggestion
    linked_attachment_ids = fields.One2many(
        'ir.attachment', 'res_id',
        domain=lambda self: [('res_model', '=', self._name)],
        readonly=True, string="Attachments",
    )
```
No need to compute, no need to add explicit invalidation in model `ir.attachment`: all those things are supported by the framework.  This solution's performance should be very close to yours.

I recommend to **not** modify attachments through this field.  There's a widget for that.

Reply:
A lot of tests won't pass with your suggestion.

> I recommend to not modify attachments through this field. There's a widget for that.

Which widget?

Reply:
Also, I am not changing anything related to ACLs, everything is done using the ORM and public methods.